### PR TITLE
fixed threadSafeMap memory leak

### PR DESCRIPTION
**Pull request by zhunzhun1988**
**What type of PR is this?**
>threadSafeMap memory leak caused by index not deleted 
>/kind bug

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
bug-fixes
```

### CODE REVIEW

**Comment by zhunzhun1988** [comment]
I signed it

**Comment by zhunzhun1988** [comment]
I signed it

**Comment by zhunzhun1988** [comment]
I signed it

**Comment by zhunzhun1988** [comment]
I signed it

File: staging/src/k8s.io/client-go/tools/cache/thread_safe_store.go [hunk 1]
```diff
@@ -292,6 +292,9 @@ func (c *threadSafeMap) deleteFromIndices(obj interface{}, key string) {
 			set := index[indexValue]
 			if set != nil {
 				set.Delete(key)
+				if len(set) == 0 {
+					delete(index, indexValue)
```
Comment (yue9944882) [inline staging/src/k8s.io/client-go/tools/cache/thread_safe_store.go]: it leaks when an empty namespace is deleted but the cache entry will retain. btw, i remember there's a duplicate pull months ago, but for now it seems leaked too.

**Comment by zhunzhun1988** [comment]
I signed it

**Comment by zhunzhun1988** [comment]
> Thanks for your pull request. Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
> 📝 **Please follow instructions at https://git.k8s.io/community/CLA.md#the-contributor-license-agreement to sign the CLA.**
> 
> It may take a couple minutes for the CLA signature to be fully registered; after that, please reply here with a new comment and we'll verify. Thanks.
> 
> * If you've already signed a CLA, it's possible we don't have your GitHub username or you're using a different email address.  Check your existing CLA data and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> * If you signed the CLA as a corporation, please sign in with your organization's credentials at https://identity.linuxfoundation.org/projects/cncf to be authorized.
> * If you have done the above and are still having issues with the CLA being reported as unsigned, please log a ticket with the Linux Foundation Helpdesk: https://support.linuxfoundation.org/
> * Should you encounter any issues with the Linux Foundation Helpdesk, send a message to the backup e-mail support address at: [login-issues@jira.linuxfoundation.org](mailto:login-issues@jira.linuxfoundation.org)
> 
> Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md). If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).

registered

**Pull request closed without merging**